src/compiler/moduleNameResolver.ts

```diff
@@ -1866,6 +1867,20 @@ function nodeModuleNameResolverWorker(
         }
     }
 
+    if (!compilerOptions.noDtsResolution && result?.value?.isExternalLibraryImport && result.value.resolved.extension === Extension.Dts) {
+        resolveModuleName(
+            moduleName,
+            combinePaths(containingDirectory, "__noDtsResolution__.ts"),
+            { ...compilerOptions, noDtsResolution: true },
+            host,
+            cache,

```

Comment:
Cache provided here does not check "options" and is suppose to be used for this resolution. 
Only project reference redirect use the key = module resolution affecting options serialized.

Reply:
FYI this experiment is just an attempt to get a ballpark idea and probably nothing like it will ever land, so I’m going to hack on this in incomplete ways